--- a/feeds/luci/applications/luci-app-package-manager/root/usr/libexec/package-manager-call
+++ b/feeds/luci/applications/luci-app-package-manager/root/usr/libexec/package-manager-call
@@ -26,7 +26,20 @@ case "$action" in
 			if [ $ipkg_bin = "apk" ]; then
 				case "$action" in
 					install)
-						action="add"
+						local_apk=
+						for arg in "$@"; do
+							case "$arg" in
+							*.apk)
+								[ -f "$arg" ] && local_apk=1 && break
+							;;
+							esac
+						done
+
+						if [ -n "$local_apk" ]; then
+							action="add --allow-untrusted"
+						else
+							action="add"
+						fi
 					;;
 					remove)
 						action="del"
