diff --git a/mt76_connac_mcu.h b/mt76_connac_mcu.h
--- a/mt76_connac_mcu.h
+++ b/mt76_connac_mcu.h
@@ -632,13 +632,6 @@
 	__le16 tag;
 	__le16 len;
 	u8 link_bitmap;
-	u8 link_ant_num[3];
-} __packed;
-
-struct sta_rec_eml_op {
-	__le16 tag;
-	__le16 len;
-	u8 bitmap;
 	u8 link_ant_num[3];
 } __packed;
 
diff --git a/mt7996/init.c b/mt7996/init.c
--- a/mt7996/init.c
+++ b/mt7996/init.c
@@ -367,26 +367,33 @@
 }
 
 static void __mt7996_init_txpower(struct mt7996_phy *phy,
-				  struct ieee80211_supported_band *sband)
+			  struct ieee80211_supported_band *sband)
 {
 	struct mt7996_dev *dev = phy->dev;
 	int i, n_chains = hweight16(phy->mt76->chainmask);
 	int path_delta = mt76_tx_power_path_delta(n_chains);
 	int pwr_delta = mt7996_eeprom_get_power_delta(dev, sband->band);
 	struct mt76_power_limits limits;
-
+	struct mt76_power_path_limits limits_path;
+	struct device_node *np;
+
+	np = mt76_find_power_limits_node(phy->mt76);
 	for (i = 0; i < sband->n_channels; i++) {
 		struct ieee80211_channel *chan = &sband->channels[i];
 		int target_power = mt7996_eeprom_get_target_power(dev, chan);
 
 		target_power += pwr_delta;
 		target_power = mt76_get_rate_power_limits(phy->mt76, chan,
-							  &limits,
-							  target_power);
+					  &limits,
+					  &limits_path,
+					  target_power);
 		target_power += path_delta;
 		target_power = DIV_ROUND_UP(target_power, 2);
-		chan->max_power = min_t(int, chan->max_reg_power,
-					target_power);
+		if (!np)
+			chan->max_power = min_t(int, chan->max_reg_power,
+						target_power);
+		else
+			chan->max_power = target_power;
 		phy->txpower = max(phy->txpower, chan->max_power);
 		chan->orig_mpwr = target_power;
 	}
@@ -394,15 +401,22 @@
 
 void mt7996_init_txpower(struct mt7996_phy *phy)
 {
+	struct mt7996_phy *phy_2g, *phy_5g, *phy_6g;
+
 	if (!phy)
 		return;
 
-	if (phy->mt76->cap.has_2ghz)
-		__mt7996_init_txpower(phy, &phy->mt76->sband_2g.sband);
-	if (phy->mt76->cap.has_5ghz)
-		__mt7996_init_txpower(phy, &phy->mt76->sband_5g.sband);
-	if (phy->mt76->cap.has_6ghz)
-		__mt7996_init_txpower(phy, &phy->mt76->sband_6g.sband);
+	/* FIXME refactor after single wiphy multiple radios merged */
+	phy_2g = mt7996_band_phy(phy->dev, NL80211_BAND_2GHZ);
+	phy_5g = mt7996_band_phy(phy->dev, NL80211_BAND_5GHZ);
+	phy_6g = mt7996_band_phy(phy->dev, NL80211_BAND_6GHZ);
+
+	if (phy_2g && phy_2g->mt76->cap.has_2ghz)
+		__mt7996_init_txpower(phy_2g, &phy_2g->mt76->sband_2g.sband);
+	if (phy_5g && phy_5g->mt76->cap.has_5ghz)
+		__mt7996_init_txpower(phy_5g, &phy_5g->mt76->sband_5g.sband);
+	if (phy_6g && phy_6g->mt76->cap.has_6ghz)
+		__mt7996_init_txpower(phy_6g, &phy_6g->mt76->sband_6g.sband);
 }
 
 static void
@@ -861,6 +875,8 @@
 
 	if (!is_mt7990(&dev->mt76))
 		mt7996_mcu_set_dup_wtbl(dev);
+
+	mt7996_low_power_config(dev);
 }
 
 void mt7996_wfsys_reset(struct mt7996_dev *dev)
@@ -1737,8 +1753,10 @@
 	if (iftype != NL80211_IFTYPE_AP)
 		return;
 
-	eht_cap_elem->mac_cap_info[1] |=
-		IEEE80211_EHT_MAC_CAP1_UNSOL_EPCS_PRIO_ACCESS;
+	if (iftype == NL80211_IFTYPE_AP)
+		eht_cap_elem->mac_cap_info[0] |=
+			IEEE80211_EHT_MAC_CAP0_SCS_TRAFFIC_DESC |
+			IEEE80211_EHT_MAC_CAP0_EPCS_PRIO_ACCESS;
 
 	eht_cap_elem->phy_cap_info[3] |=
 		IEEE80211_EHT_PHY_CAP3_TRIG_SU_BF_FDBK |
diff --git a/mt7996/mac.c b/mt7996/mac.c
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -3048,7 +3048,7 @@
 	if (en)
 		ieee80211_hw_set(hw, SUPPORTS_AMSDU_IN_AMPDU);
 	else
-		ieee80211_hw_clear(hw, SUPPORTS_AMSDU_IN_AMPDU);
+		__clear_bit(IEEE80211_HW_SUPPORTS_AMSDU_IN_AMPDU, hw->flags);
 }
 
 void mt7996_mac_sta_rc_work(struct work_struct *work)
diff --git a/mt7996/main.c b/mt7996/main.c
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -2836,21 +2836,6 @@
 	mt7996_for_each_phy(dev, phy)
 		ieee80211_queue_delayed_work(hw, &phy->mt76->mac_work,
 					     MT7996_WATCHDOG_TIME);
-}
-
-static int
-mt7996_set_eml_op_mode(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
-		       struct ieee80211_sta *sta,
-		       struct ieee80211_eml_params *eml_params)
-{
-	struct mt7996_dev *dev = mt7996_hw_dev(hw);
-	int ret;
-
-	mutex_lock(&dev->mt76.mutex);
-	ret = mt7996_mcu_set_emlsr_mode(dev, vif, sta, eml_params);
-	mutex_unlock(&dev->mt76.mutex);
-
-	return ret;
 }
 
 static int
@@ -2985,5 +2970,4 @@
 	.change_vif_links = mt7996_change_vif_links,
 	.change_sta_links = mt7996_mac_sta_change_links,
 	.reconfig_complete = mt7996_reconfig_complete,
-	.set_eml_op_mode = mt7996_set_eml_op_mode,
 };
diff --git a/mt7996/mcu.c b/mt7996/mcu.c
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -811,8 +811,7 @@
 	struct ieee80211_vif *vif;
 	struct wireless_dev *wdev;
 
-	if (wiphy_ext_feature_isset(mphy->hw->wiphy,
-				    NL80211_EXT_FEATURE_STAS_COUNT)) {
+	if (false) {
 		msta_link = container_of(wcid, struct mt7996_sta_link, wcid);
 		vif = container_of((void *)msta_link->sta->vif, struct ieee80211_vif,
 				   drv_priv);
@@ -822,7 +821,7 @@
 			return;
 
 		dev_sw_netstats_tx_add(wdev->netdev, tx_packets, tx_bytes);
-		__dev_sw_netstats_rx_add(wdev->netdev, rx_packets, rx_bytes);
+		dev_sw_netstats_rx_add(wdev->netdev, rx_bytes);
 	}
 }
 
@@ -2233,61 +2232,6 @@
 				     MCU_WM_UNI_CMD(BSS_INFO_UPDATE), true);
 }
 
-int mt7996_mcu_set_emlsr_mode(struct mt7996_dev *dev,
-			      struct ieee80211_vif *vif,
-			      struct ieee80211_sta *sta,
-			      struct ieee80211_eml_params *eml_params)
-{
-	struct mt7996_sta *msta = (struct mt7996_sta *)sta->drv_priv;
-	struct mt7996_sta_link *msta_link;
-	struct sta_rec_eml_op *eml_op;
-	struct mt7996_vif_link *link;
-	struct sk_buff *skb;
-	struct tlv *tlv;
-
-	msta_link = mt76_dereference(msta->link[eml_params->link_id],
-				     &dev->mt76);
-	if (!msta_link)
-		return -EINVAL;
-
-	link = mt7996_vif_link(dev, vif, eml_params->link_id);
-	if (!link)
-		return -EINVAL;
-
-	skb = __mt76_connac_mcu_alloc_sta_req(&dev->mt76, &link->mt76,
-					      &msta_link->wcid,
-					      MT7996_STA_UPDATE_MAX_SIZE);
-	if (IS_ERR(skb))
-		return PTR_ERR(skb);
-
-	tlv = mt76_connac_mcu_add_tlv(skb, STA_REC_EML_OP, sizeof(*eml_op));
-	eml_op = (struct sta_rec_eml_op *)tlv;
-	eml_op->link_bitmap = 0;
-
-	if (eml_params->control & IEEE80211_EML_CTRL_EMLSR_MODE) {
-		unsigned long link_bitmap = eml_params->link_bitmap;
-		unsigned int link_id;
-
-		for_each_set_bit(link_id, &link_bitmap,
-				 IEEE80211_MLD_MAX_NUM_LINKS) {
-			struct mt76_phy *mphy;
-
-			link = mt7996_vif_link(dev, vif, link_id);
-			if (!link)
-				continue;
-
-			mphy = mt76_vif_link_phy(&link->mt76);
-			if (!mphy)
-				continue;
-
-			eml_op->link_bitmap |= BIT(mphy->band_idx);
-		}
-	}
-
-	return mt76_mcu_skb_send_msg(&dev->mt76, skb,
-				     MCU_WMWA_UNI_CMD(STA_REC_UPDATE), true);
-}
-
 int mt7996_mcu_get_tsf_offset(struct mt7996_phy *phy,
 			      struct mt7996_vif *mvif,
 			      int rpting_link_id,
@@ -3766,7 +3710,7 @@
 	for (i = 0; i < ARRAY_SIZE(eht_mld->str_cap); i++)
 		eht_mld->str_cap[i] = 0x7;
 
-	eht_mld->eml_cap = cpu_to_le16(sta->eml_capa);
+	eht_mld->eml_cap = 0;
 }
 
 int mt7996_mcu_add_sta(struct mt7996_dev *dev, struct ieee80211_vif *vif,
diff --git a/mt7996/mcu.h b/mt7996/mcu.h
--- a/mt7996/mcu.h
+++ b/mt7996/mcu.h
@@ -204,16 +204,6 @@
 	__le32 type;
 	__le32 rsv[4];
 	union eeprom_data eeprom;
-} __packed;
-
-struct mt7996_mcu_eeprom_info {
-	u8 _rsv[4];
-
-	__le16 tag;
-	__le16 len;
-	__le32 addr;
-	__le32 valid;
-	u8 data[MT7996_EEPROM_BLOCK_SIZE];
 } __packed;
 
 struct mt7996_mcu_eeprom_patch {
@@ -904,28 +894,6 @@
 	u8 bss_idx;
 	u8 __rsv;
 } __packed;
-
-struct mld_req_hdr {
-	u8 ver;
-	u8 mld_addr[ETH_ALEN];
-	u8 mld_idx;
-	u8 flag;
-	u8 rsv[3];
-	u8 buf[];
-} __packed;
-
-struct mld_reconf_stop_link {
-	__le16 tag;
-	__le16 len;
-	__le16 link_bitmap;
-	u8 rsv[2];
-	u8 bss_idx[16];
-} __packed;
-
-enum {
-	UNI_CMD_MLD_RECONF_AP_REM_TIMER = 0x03,
-	UNI_CMD_MLD_RECONF_STOP_LINK = 0x04,
-};
 
 struct hdr_trans_en {
 	__le16 tag;
diff --git a/mt7996/mt7996.h b/mt7996/mt7996.h
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -50,26 +50,31 @@
 #define MT7996_FIRMWARE_WA		"mediatek/mt7996/mt7996_wa.bin"
 #define MT7996_FIRMWARE_WM		"mediatek/mt7996/mt7996_wm.bin"
 #define MT7996_FIRMWARE_DSP		"mediatek/mt7996/mt7996_dsp.bin"
+#define MT7996_FIRMWARE_WM_TM		"mediatek/mt7996/mt7996_wm_tm.bin"
 #define MT7996_ROM_PATCH		"mediatek/mt7996/mt7996_rom_patch.bin"
 
 #define MT7996_FIRMWARE_WA_233		"mediatek/mt7996/mt7996_wa_233.bin"
 #define MT7996_FIRMWARE_WM_233		"mediatek/mt7996/mt7996_wm_233.bin"
 #define MT7996_FIRMWARE_DSP_233		MT7996_FIRMWARE_DSP
+#define MT7996_FIRMWARE_WM_TM_233	"mediatek/mt7996/mt7996_wm_tm_233.bin"
 #define MT7996_ROM_PATCH_233		"mediatek/mt7996/mt7996_rom_patch_233.bin"
 
 #define MT7992_FIRMWARE_WA		"mediatek/mt7996/mt7992_wa.bin"
 #define MT7992_FIRMWARE_WM		"mediatek/mt7996/mt7992_wm.bin"
 #define MT7992_FIRMWARE_DSP		"mediatek/mt7996/mt7992_dsp.bin"
+#define MT7992_FIRMWARE_WM_TM		"mediatek/mt7996/mt7992_wm_tm.bin"
 #define MT7992_ROM_PATCH		"mediatek/mt7996/mt7992_rom_patch.bin"
 
 #define MT7992_FIRMWARE_WA_23		"mediatek/mt7996/mt7992_wa_23.bin"
 #define MT7992_FIRMWARE_WM_23		"mediatek/mt7996/mt7992_wm_23.bin"
 #define MT7992_FIRMWARE_DSP_23		"mediatek/mt7996/mt7992_dsp_23.bin"
+#define MT7992_FIRMWARE_WM_TM_23	"mediatek/mt7996/mt7992_wm_tm_23.bin"
 #define MT7992_ROM_PATCH_23		"mediatek/mt7996/mt7992_rom_patch_23.bin"
 
 #define MT7992_FIRMWARE_WA_24		"mediatek/mt7996/mt7992_wa_24.bin"
 #define MT7992_FIRMWARE_WM_24		"mediatek/mt7996/mt7992_wm_24.bin"
 #define MT7992_FIRMWARE_DSP_24		"mediatek/mt7996/mt7992_dsp_24.bin"
+#define MT7992_FIRMWARE_WM_TM_24	"mediatek/mt7996/mt7992_wm_tm_24.bin"
 #define MT7992_ROM_PATCH_24		"mediatek/mt7996/mt7992_rom_patch_24.bin"
 
 #define MT7990_FIRMWARE_WA		""
@@ -82,6 +87,7 @@
 #define MT7996_EEPROM_DEFAULT_INT	"mediatek/mt7996/mt7996_eeprom_2i5i6i.bin"
 #define MT7996_EEPROM_DEFAULT_233	"mediatek/mt7996/mt7996_eeprom_233.bin"
 #define MT7996_EEPROM_DEFAULT_233_INT	"mediatek/mt7996/mt7996_eeprom_233_2i5i6i.bin"
+#define MT7996_EEPROM_DEFAULT_404	"mediatek/mt7996/mt7996_eeprom_dual_404.bin"
 
 #define MT7992_EEPROM_DEFAULT		"mediatek/mt7996/mt7992_eeprom.bin"
 #define MT7992_EEPROM_DEFAULT_INT	"mediatek/mt7996/mt7992_eeprom_2i5i.bin"
@@ -97,6 +103,7 @@
 #define MT7996_EEPROM_BLOCK_SIZE	16
 #define MT7996_EXT_EEPROM_BLOCK_SIZE	1024
 #define MT7996_TOKEN_SIZE		16384
+#define MT7996_WED_TOKEN_SIZE		16384
 #define MT7996_HW_TOKEN_SIZE		8192
 #define MT7996_SW_TOKEN_SIZE		15360
 #define MT7996_PER_BAND_TOKEN_SIZE	4000
@@ -128,6 +135,8 @@
 #define MT7996_MAX_TEMP_IDX		1
 #define MT7996_CRIT_TEMP		110
 #define MT7996_MAX_TEMP			120
+
+#define MT7996_BUILD_TIME_LEN		24
 
 #define MT7996_MAX_HIF_RXD_IN_PG	5
 #define MT7996_RRO_MSDU_PG_HASH_SIZE	127
@@ -170,6 +179,7 @@
 #define IP_DSCP_NUM			64
 #define ADDBA_RETRY_PERIOD		(5 * HZ)
 
+struct mt7996_dev;
 struct mt7996_vif;
 struct mt7996_sta;
 struct mt7996_dfs_pulse;
@@ -177,13 +187,16 @@
 
 enum mt7996_ram_type {
 	MT7996_RAM_TYPE_WM,
+	MT7996_RAM_TYPE_WM_TM = MT7996_RAM_TYPE_WM,
 	MT7996_RAM_TYPE_WA,
 	MT7996_RAM_TYPE_DSP,
+	__MT7996_RAM_TYPE_MAX,
 };
 
 enum mt7996_var_type {
 	MT7996_VAR_TYPE_444,
 	MT7996_VAR_TYPE_233,
+	MT7996_VAR_TYPE_404,
 };
 
 enum mt7992_var_type {
@@ -1340,10 +1353,9 @@
 int mt7996_mcu_ps_leave(struct mt7996_dev *dev, struct mt7996_vif_link *mconf,
 			struct mt7996_sta_link *msta_link);
 int mt7996_mcu_cp_support(struct mt7996_dev *dev, u8 mode);
-int mt7996_mcu_set_emlsr_mode(struct mt7996_dev *dev,
-			      struct ieee80211_vif *vif,
-			      struct ieee80211_sta *sta,
-			      struct ieee80211_eml_params *eml_params);
+int mt7996_mcu_set_tpo(struct mt7996_dev *dev, u8 type, u8 val);
+int mt7996_mcu_set_lp_option(struct mt7996_dev *dev, u8 *arg);
+int mt7996_mcu_set_pst(struct mt7996_dev *dev, u32 band, u32 cmd, u32 val);
 #ifdef CONFIG_MAC80211_DEBUGFS
 void mt7996_sta_add_debugfs(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 			    struct ieee80211_sta *sta, struct dentry *dir);
@@ -760,6 +760,15 @@ struct mt7996_dev {
 
 	u8 eeprom_mode;
 	bool has_eht:1;
+	bool flash_mode:1;
+	bool bin_file_mode:1;
+	u32 cur_prek_offset;
+	u8 *cal;
+	const char *bin_file_name;
+	struct {
+		const u32 *rev;
+		u32 dpd_ch_num[DPD_CH_NUM_TYPE_MAX];
+	} prek;
 
 	struct {
 		struct {
 			void *ptr;
diff --git a/mt7996/eeprom.c b/mt7996/eeprom.c
--- a/mt7996/eeprom.c
+++ b/mt7996/eeprom.c
@@ -301,6 +301,10 @@ static int mt7996_eeprom_load(struct mt7996_dev *dev)
 	}
 
 out:
+	dev->flash_mode = dev->eeprom_mode == EEPROM_MODE_FLASH ||
+			  dev->eeprom_mode == EEPROM_MODE_EXT;
+	dev->bin_file_mode = false;
+	dev->bin_file_name = NULL;
 	return mt7996_eeprom_check_or_use_default(dev, use_default);
 }
 
@@ -422,7 +426,7 @@ mt7996_eeprom_load_precal_binfile(struct mt7996_dev *dev, u32 offs, u32 size)
 	const struct firmware *fw = NULL;
 	int ret;
 
-	ret = request_firmware(&fw, dev->mt76.bin_file_name, dev->mt76.dev);
+	ret = request_firmware(&fw, dev->bin_file_name, dev->mt76.dev);
 	if (ret)
 		return ret;
 
diff --git a/mt7996/mtk_debugfs.c b/mt7996/mtk_debugfs.c
--- a/mt7996/mtk_debugfs.c
+++ b/mt7996/mtk_debugfs.c
@@ -2931,7 +2931,7 @@ static int mt7996_show_eeprom_mode(struct seq_file *s, void *data)
 			seq_printf(s, "   flash mode\n");
 		break;
 	case BIN_FILE_MODE:
-		seq_printf(s, "   bin file mode\n   filename = %s\n", dev->mt76.bin_file_name);
+		seq_printf(s, "   bin file mode\n   filename = %s\n", dev->bin_file_name ?: "(unset)");
 		break;
 	default:
 		break;
diff --git a/mt7996/pci.c b/mt7996/pci.c
--- a/mt7996/pci.c
+++ b/mt7996/pci.c
@@ -179,10 +179,11 @@
 
 	if (hif2) {
 		hif2_dev = container_of(hif2->dev, struct pci_dev, dev);
-
-		ret = mt7996_mmio_wed_init(dev, hif2_dev, true, &hif2_irq);
+		ret = 0;
+
+		ret = mt7996_mmio_wed_init(dev, hif2_dev, true, &irq);
 		if (ret < 0)
-			goto free_hif2_wed_irq_vector;
+			goto free_wed_or_irq_vector;
 
 		if (!ret) {
 			ret = pci_alloc_irq_vectors(hif2_dev, 1, 1,
