diff --git a/mt7996/mac.c b/mt7996/mac.c
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -3048,7 +3048,7 @@
 	if (en)
 		ieee80211_hw_set(hw, SUPPORTS_AMSDU_IN_AMPDU);
 	else
-		ieee80211_hw_clear(hw, SUPPORTS_AMSDU_IN_AMPDU);
+		__clear_bit(IEEE80211_HW_SUPPORTS_AMSDU_IN_AMPDU, hw->flags);
 }
 
 void mt7996_mac_sta_rc_work(struct work_struct *work)
diff --git a/mt7996/main.c b/mt7996/main.c
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -2836,21 +2836,6 @@
 	mt7996_for_each_phy(dev, phy)
 		ieee80211_queue_delayed_work(hw, &phy->mt76->mac_work,
 					     MT7996_WATCHDOG_TIME);
-}
-
-static int
-mt7996_set_eml_op_mode(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
-		       struct ieee80211_sta *sta,
-		       struct ieee80211_eml_params *eml_params)
-{
-	struct mt7996_dev *dev = mt7996_hw_dev(hw);
-	int ret;
-
-	mutex_lock(&dev->mt76.mutex);
-	ret = mt7996_mcu_set_emlsr_mode(dev, vif, sta, eml_params);
-	mutex_unlock(&dev->mt76.mutex);
-
-	return ret;
 }
 
 static int
@@ -2985,5 +2970,4 @@
 	.change_vif_links = mt7996_change_vif_links,
 	.change_sta_links = mt7996_mac_sta_change_links,
 	.reconfig_complete = mt7996_reconfig_complete,
-	.set_eml_op_mode = mt7996_set_eml_op_mode,
 };
diff --git a/mt7996/mcu.c b/mt7996/mcu.c
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -2233,61 +2233,6 @@
 				     MCU_WM_UNI_CMD(BSS_INFO_UPDATE), true);
 }
 
-int mt7996_mcu_set_emlsr_mode(struct mt7996_dev *dev,
-			      struct ieee80211_vif *vif,
-			      struct ieee80211_sta *sta,
-			      struct ieee80211_eml_params *eml_params)
-{
-	struct mt7996_sta *msta = (struct mt7996_sta *)sta->drv_priv;
-	struct mt7996_sta_link *msta_link;
-	struct sta_rec_eml_op *eml_op;
-	struct mt7996_vif_link *link;
-	struct sk_buff *skb;
-	struct tlv *tlv;
-
-	msta_link = mt76_dereference(msta->link[eml_params->link_id],
-				     &dev->mt76);
-	if (!msta_link)
-		return -EINVAL;
-
-	link = mt7996_vif_link(dev, vif, eml_params->link_id);
-	if (!link)
-		return -EINVAL;
-
-	skb = __mt76_connac_mcu_alloc_sta_req(&dev->mt76, &link->mt76,
-					      &msta_link->wcid,
-					      MT7996_STA_UPDATE_MAX_SIZE);
-	if (IS_ERR(skb))
-		return PTR_ERR(skb);
-
-	tlv = mt76_connac_mcu_add_tlv(skb, STA_REC_EML_OP, sizeof(*eml_op));
-	eml_op = (struct sta_rec_eml_op *)tlv;
-	eml_op->link_bitmap = 0;
-
-	if (eml_params->control & IEEE80211_EML_CTRL_EMLSR_MODE) {
-		unsigned long link_bitmap = eml_params->link_bitmap;
-		unsigned int link_id;
-
-		for_each_set_bit(link_id, &link_bitmap,
-				 IEEE80211_MLD_MAX_NUM_LINKS) {
-			struct mt76_phy *mphy;
-
-			link = mt7996_vif_link(dev, vif, link_id);
-			if (!link)
-				continue;
-
-			mphy = mt76_vif_link_phy(&link->mt76);
-			if (!mphy)
-				continue;
-
-			eml_op->link_bitmap |= BIT(mphy->band_idx);
-		}
-	}
-
-	return mt76_mcu_skb_send_msg(&dev->mt76, skb,
-				     MCU_WMWA_UNI_CMD(STA_REC_UPDATE), true);
-}
-
 int mt7996_mcu_get_tsf_offset(struct mt7996_phy *phy,
 			      struct mt7996_vif *mvif,
 			      int rpting_link_id,
@@ -3766,7 +3711,7 @@
 	for (i = 0; i < ARRAY_SIZE(eht_mld->str_cap); i++)
 		eht_mld->str_cap[i] = 0x7;
 
-	eht_mld->eml_cap = cpu_to_le16(sta->eml_capa);
+	eht_mld->eml_cap = 0;
 }
 
 int mt7996_mcu_add_sta(struct mt7996_dev *dev, struct ieee80211_vif *vif,
diff --git a/mt7996/mt7996.h b/mt7996/mt7996.h
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -50,26 +50,31 @@
 #define MT7996_FIRMWARE_WA		"mediatek/mt7996/mt7996_wa.bin"
 #define MT7996_FIRMWARE_WM		"mediatek/mt7996/mt7996_wm.bin"
 #define MT7996_FIRMWARE_DSP		"mediatek/mt7996/mt7996_dsp.bin"
+#define MT7996_FIRMWARE_WM_TM		"mediatek/mt7996/mt7996_wm_tm.bin"
 #define MT7996_ROM_PATCH		"mediatek/mt7996/mt7996_rom_patch.bin"
 
 #define MT7996_FIRMWARE_WA_233		"mediatek/mt7996/mt7996_wa_233.bin"
 #define MT7996_FIRMWARE_WM_233		"mediatek/mt7996/mt7996_wm_233.bin"
 #define MT7996_FIRMWARE_DSP_233		MT7996_FIRMWARE_DSP
+#define MT7996_FIRMWARE_WM_TM_233	"mediatek/mt7996/mt7996_wm_tm_233.bin"
 #define MT7996_ROM_PATCH_233		"mediatek/mt7996/mt7996_rom_patch_233.bin"
 
 #define MT7992_FIRMWARE_WA		"mediatek/mt7996/mt7992_wa.bin"
 #define MT7992_FIRMWARE_WM		"mediatek/mt7996/mt7992_wm.bin"
 #define MT7992_FIRMWARE_DSP		"mediatek/mt7996/mt7992_dsp.bin"
+#define MT7992_FIRMWARE_WM_TM		"mediatek/mt7996/mt7992_wm_tm.bin"
 #define MT7992_ROM_PATCH		"mediatek/mt7996/mt7992_rom_patch.bin"
 
 #define MT7992_FIRMWARE_WA_23		"mediatek/mt7996/mt7992_wa_23.bin"
 #define MT7992_FIRMWARE_WM_23		"mediatek/mt7996/mt7992_wm_23.bin"
 #define MT7992_FIRMWARE_DSP_23		"mediatek/mt7996/mt7992_dsp_23.bin"
+#define MT7992_FIRMWARE_WM_TM_23	"mediatek/mt7996/mt7992_wm_tm_23.bin"
 #define MT7992_ROM_PATCH_23		"mediatek/mt7996/mt7992_rom_patch_23.bin"
 
 #define MT7992_FIRMWARE_WA_24		"mediatek/mt7996/mt7992_wa_24.bin"
 #define MT7992_FIRMWARE_WM_24		"mediatek/mt7996/mt7992_wm_24.bin"
 #define MT7992_FIRMWARE_DSP_24		"mediatek/mt7996/mt7992_dsp_24.bin"
+#define MT7992_FIRMWARE_WM_TM_24	"mediatek/mt7996/mt7992_wm_tm_24.bin"
 #define MT7992_ROM_PATCH_24		"mediatek/mt7996/mt7992_rom_patch_24.bin"
 
 #define MT7990_FIRMWARE_WA		""
@@ -82,6 +87,7 @@
 #define MT7996_EEPROM_DEFAULT_INT	"mediatek/mt7996/mt7996_eeprom_2i5i6i.bin"
 #define MT7996_EEPROM_DEFAULT_233	"mediatek/mt7996/mt7996_eeprom_233.bin"
 #define MT7996_EEPROM_DEFAULT_233_INT	"mediatek/mt7996/mt7996_eeprom_233_2i5i6i.bin"
+#define MT7996_EEPROM_DEFAULT_404	"mediatek/mt7996/mt7996_eeprom_dual_404.bin"
 
 #define MT7992_EEPROM_DEFAULT		"mediatek/mt7996/mt7992_eeprom.bin"
 #define MT7992_EEPROM_DEFAULT_INT	"mediatek/mt7996/mt7992_eeprom_2i5i.bin"
@@ -97,6 +103,7 @@
 #define MT7996_EEPROM_BLOCK_SIZE	16
 #define MT7996_EXT_EEPROM_BLOCK_SIZE	1024
 #define MT7996_TOKEN_SIZE		16384
+#define MT7996_WED_TOKEN_SIZE		16384
 #define MT7996_HW_TOKEN_SIZE		8192
 #define MT7996_SW_TOKEN_SIZE		15360
 #define MT7996_PER_BAND_TOKEN_SIZE	4000
@@ -128,6 +135,8 @@
 #define MT7996_MAX_TEMP_IDX		1
 #define MT7996_CRIT_TEMP		110
 #define MT7996_MAX_TEMP			120
+
+#define MT7996_BUILD_TIME_LEN		24
 
 #define MT7996_MAX_HIF_RXD_IN_PG	5
 #define MT7996_RRO_MSDU_PG_HASH_SIZE	127
@@ -170,6 +179,7 @@
 #define IP_DSCP_NUM			64
 #define ADDBA_RETRY_PERIOD		(5 * HZ)
 
+struct mt7996_dev;
 struct mt7996_vif;
 struct mt7996_sta;
 struct mt7996_dfs_pulse;
@@ -177,13 +187,16 @@
 
 enum mt7996_ram_type {
 	MT7996_RAM_TYPE_WM,
+	MT7996_RAM_TYPE_WM_TM = MT7996_RAM_TYPE_WM,
 	MT7996_RAM_TYPE_WA,
 	MT7996_RAM_TYPE_DSP,
+	__MT7996_RAM_TYPE_MAX,
 };
 
 enum mt7996_var_type {
 	MT7996_VAR_TYPE_444,
 	MT7996_VAR_TYPE_233,
+	MT7996_VAR_TYPE_404,
 };
 
 enum mt7992_var_type {
@@ -1340,10 +1353,6 @@
 int mt7996_mcu_ps_leave(struct mt7996_dev *dev, struct mt7996_vif_link *mconf,
 			struct mt7996_sta_link *msta_link);
 int mt7996_mcu_cp_support(struct mt7996_dev *dev, u8 mode);
-int mt7996_mcu_set_emlsr_mode(struct mt7996_dev *dev,
-			      struct ieee80211_vif *vif,
-			      struct ieee80211_sta *sta,
-			      struct ieee80211_eml_params *eml_params);
 #ifdef CONFIG_MAC80211_DEBUGFS
 void mt7996_sta_add_debugfs(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 			    struct ieee80211_sta *sta, struct dentry *dir);
diff --git a/mt7996/pci.c b/mt7996/pci.c
--- a/mt7996/pci.c
+++ b/mt7996/pci.c
@@ -179,10 +179,11 @@
 
 	if (hif2) {
 		hif2_dev = container_of(hif2->dev, struct pci_dev, dev);
-
-		ret = mt7996_mmio_wed_init(dev, hif2_dev, true, &hif2_irq);
+		ret = 0;
+
+		ret = mt7996_mmio_wed_init(dev, hif2_dev, true, &irq);
 		if (ret < 0)
-			goto free_hif2_wed_irq_vector;
+			goto free_wed_or_irq_vector;
 
 		if (!ret) {
 			ret = pci_alloc_irq_vectors(hif2_dev, 1, 1,
