diff --git a/feeds/luci/modules/luci-mod-status/htdocs/luci-static/resources/view/status/include/60_wifi.js b/feeds/luci/modules/luci-mod-status/htdocs/luci-static/resources/view/status/include/60_wifi.js
index 95f8368..af95283 100644
--- a/feeds/luci/modules/luci-mod-status/htdocs/luci-static/resources/view/status/include/60_wifi.js
+++ b/feeds/luci/modules/luci-mod-status/htdocs/luci-static/resources/view/status/include/60_wifi.js
@@ -37,6 +37,7 @@ return baseclass.extend({
 
 		var s = '%.1f\xa0%s'.format(rt.rate / 1000, _('Mbit/s')),
 		    ht = rt.ht, vht = rt.vht,
+			mhz = (rt.mhz != null) ? ((rt.mhz_hi && rt.mhz < 256) ? ((rt.mhz_hi << 8) + rt.mhz) : rt.mhz) : null,
 			nss = rt.nss,
 			mcs = rt.mcs, sgi = rt.short_gi,
 			he = rt.he, he_gi = rt.he_gi,
@@ -45,9 +46,9 @@ return baseclass.extend({
 			eht_gi = rt.eht_gi,
 			eht_dcm = rt.eht_dcm;
 
-		if (rt.mhz != null)
-			s += ', %d\xa0%s'.format(rt.mhz, _('MHz'));
+		if (mhz != null)
+			s += ', %d\xa0%s'.format(mhz, _('MHz'));
 
 		if (ht || vht) {
 			if (vht && mcs != null) s += ', VHT-MCS\xa0%d'.format(mcs);
