From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: diy-mtk.sh <build@localhost>
Date: Mon, 14 Jul 2026 00:00:00 +0000
Subject: [PATCH] net: ethernet: mtk_eth_soc: add mtkhnat Kconfig

---
 drivers/net/ethernet/mediatek/Kconfig | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

--- a/drivers/net/ethernet/mediatek/Kconfig
+++ b/drivers/net/ethernet/mediatek/Kconfig
@@ -42,2 +42,26 @@ config NET_MEDIATEK_STAR_EMAC
 	  This driver supports the ethernet MAC IP first used on
 	  MediaTek MT85** SoCs.
+
+config MEDIATEK_NETSYS_V2
+	tristate "MediaTek Ethernet NETSYS V2 support"
+	depends on ARCH_MEDIATEK && NET_MEDIATEK_SOC
+	help
+	  This options enable MTK Ethernet NETSYS V2 support
+
+config MEDIATEK_NETSYS_V3
+	tristate "MediaTek Ethernet NETSYS V3 support"
+	depends on ARCH_MEDIATEK && NET_MEDIATEK_SOC
+	help
+	  This options enable MTK Ethernet NETSYS V3 support for
+	  XGMAC and USXGMII.
+
+config NET_MEDIATEK_HNAT
+	tristate "MediaTek HW NAT support"
+	depends on NET_MEDIATEK_SOC && NF_CONNTRACK && IP_NF_NAT
+	help
+	  This driver supports the Hardware Network Address Translation
+	  in the MediaTek MT7988/MT7986/MT7981 chipset family.
