--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -2233,6 +2233,13 @@ next_frag:
 	state.tx_buf->type = MTK_TYPE_SKB;
 	state.tx_buf->data = skb;
 
+#if defined(CONFIG_NET_MEDIATEK_HNAT) || defined(CONFIG_NET_MEDIATEK_HNAT_MODULE)
+	if (HNAT_SKB_CB2(skb)->magic == 0x78681415) {
+		itxd->txd4 &= ~(0xf << TX_DMA_FPORT_SHIFT_V2);
+		itxd->txd4 |= 0x3 << TX_DMA_FPORT_SHIFT_V2;
+	}
+#endif
+
 	netdev_tx_sent_queue(txq, skb->len);
 	skb_tx_timestamp(skb);
 
