--- a/drivers/net/ethernet/mediatek/mtk_hnat/hnat_nf_hook.c
+++ b/drivers/net/ethernet/mediatek/mtk_hnat/hnat_nf_hook.c
@@ -460,12 +460,19 @@ int nf_hnat_netdevice_event(struct notif
 			dev_put(dev);
 		}
 
+		if (g_rx_ppdev == dev) {
+			WRITE_ONCE(g_rx_ppdev, NULL);
+			dev_put(dev);
+		}
+
 		break;
 	case NETDEV_REGISTER:
 		if (IS_PPD(dev) && !hnat_priv->g_ppdev)
 			hnat_priv->g_ppdev = dev_get_by_name(&init_net, hnat_priv->ppd);
 		if (IS_WAN(dev) && !hnat_priv->g_wandev)
 			hnat_priv->g_wandev = dev_get_by_name(&init_net, hnat_priv->wan);
+		if (IS_RX_PPD(dev) && !g_rx_ppdev)
+			g_rx_ppdev = dev_get_by_name(&init_net, hnat_priv->rx_ppd);
 		break;
 	case MTK_FE_RESET_NAT_DONE:
 		pr_info("[%s] HNAT driver starts to do warm init !\n", __func__);
--- a/drivers/net/ethernet/mediatek/mtk_hnat/hnat.h
+++ b/drivers/net/ethernet/mediatek/mtk_hnat/hnat.h
@@ -1060,6 +1060,7 @@ struct mtk_hnat {
 	char lan[IFNAMSIZ];
 	char lan2[IFNAMSIZ];
 	char ppd[IFNAMSIZ];
+	char rx_ppd[IFNAMSIZ];
 	u16 lvid;
 	u16 wvid;
 
@@ -1343,6 +1344,7 @@ enum FoeIpAct {
 	 (get_wifi_hook_if_index_from_dev(dev) != 0)) ? 1 : 0)
 #define IS_EXT(dev) ((get_index_from_dev(dev) != 0) ? 1 : 0)
 #define IS_PPD(dev) (!strcmp(dev->name, hnat_priv->ppd))
+#define IS_RX_PPD(dev) (!strcmp(dev->name, hnat_priv->rx_ppd))
 #define IS_L2_BRIDGE(x) (((x)->bfib1.pkt_type == L2_BRIDGE) ? 1 : 0)
 #define IS_IPV4_HNAPT(x) (((x)->bfib1.pkt_type == IPV4_HNAPT) ? 1 : 0)
 #define IS_IPV4_HNAT(x) (((x)->bfib1.pkt_type == IPV4_HNAT) ? 1 : 0)
@@ -1439,6 +1441,8 @@ enum FoeIpAct {
 extern const struct of_device_id of_hnat_match[];
 extern struct mtk_hnat *hnat_priv;
 
+extern struct net_device *g_rx_ppdev;
+
 int hnat_dsa_fill_stag(const struct net_device *netdev,
 		       struct foe_entry *entry,
 		       u16 eth_proto, int mape);
--- a/drivers/net/ethernet/mediatek/mtk_hnat/hnat.c
+++ b/drivers/net/ethernet/mediatek/mtk_hnat/hnat.c
@@ -1397,6 +1397,9 @@ static void hnat_release_netdev(void)
 
 	if (hnat_priv->g_wandev)
 		dev_put(hnat_priv->g_wandev);
+
+	if (g_rx_ppdev)
+		dev_put(g_rx_ppdev);
 }
 
 static struct notifier_block nf_hnat_netdevice_nb __read_mostly = {
@@ -1608,6 +1611,13 @@ static int hnat_probe(struct platform_de
 		strncpy(hnat_priv->ppd, (char *)name, IFNAMSIZ - 1);
 	dev_info(&pdev->dev, "ppd = %s\n", hnat_priv->ppd);
 
+	err = of_property_read_string(np, "mtketh-rx-ppd", &name);
+	if (err < 0)
+		strscpy(hnat_priv->rx_ppd, "rxppd", IFNAMSIZ);
+	else
+		strscpy(hnat_priv->rx_ppd, name, IFNAMSIZ);
+	dev_info(&pdev->dev, "rx-ppd = %s\n", hnat_priv->rx_ppd);
+
 	/*get total gmac num in hnat*/
 	err = of_property_read_u32_index(np, "mtketh-max-gmac", 0, &val);
 
