diff --git a/mt7996/main.c b/mt7996/main.c
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -765,11 +765,23 @@
 	if (mphy->offchannel)
 		mt7996_mac_update_beacons(phy);
 
-	ret = mt7996_mcu_set_chan_info(phy, UNI_CHANNEL_SWITCH);
+
+	if (phy->dev->cal) {
+		ret = mt7996_mcu_apply_tx_dpd(phy);
+		if (ret)
+			goto out;
+	}
+
+	if (mt76_testmode_enabled(phy->mt76) || phy->mt76->test.bf_en) {
+		mt7996_tm_update_channel(phy);
+		goto out;
+	}
+
+	ret = mt7996_mcu_set_chan_info(phy, UNI_CHANNEL_SWITCH, false);
 	if (ret)
 		goto out;
 
-	ret = mt7996_mcu_set_chan_info(phy, UNI_CHANNEL_RX_PATH);
+	ret = mt7996_mcu_set_chan_info(phy, UNI_CHANNEL_RX_PATH, false);
 	if (ret)
 		goto out;
 
diff --git a/mt7996/mcu.c b/mt7996/mcu.c
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -4692,7 +4692,7 @@
 	return mt7996_mcu_rdd_cmd(dev, RDD_START, rdd_idx, region);
 }
 
-int mt7996_mcu_set_chan_info(struct mt7996_phy *phy, u16 tag)
+int mt7996_mcu_set_chan_info(struct mt7996_phy *phy, u16 tag, bool sta)
 {
 	static const u8 ch_band[] = {
 		[NL80211_BAND_2GHZ] = 0,
@@ -4703,6 +4703,8 @@
 	struct cfg80211_chan_def *chandef = &phy->mt76->chandef;
 	int freq1 = chandef->center_freq1;
 	u8 band_idx = phy->mt76->band_idx;
+	enum nl80211_iftype iftype = sta ? NL80211_IFTYPE_STATION :
+					   NL80211_IFTYPE_AP;
 	struct {
 		/* fixed field */
 		u8 __rsv[4];
@@ -4741,8 +4743,7 @@
 		req.switch_reason = CH_SWITCH_NORMAL;
 	else if (phy->mt76->offchannel || !phy->mt76->chanctx)
 		req.switch_reason = CH_SWITCH_SCAN_BYPASS_DPD;
-	else if (!cfg80211_reg_can_beacon(phy->mt76->hw->wiphy, chandef,
-					  NL80211_IFTYPE_AP))
+	else if (!cfg80211_reg_can_beacon(phy->mt76->hw->wiphy, chandef, iftype))
 		req.switch_reason = CH_SWITCH_DFS;
 	else
 		req.switch_reason = CH_SWITCH_NORMAL;
diff --git a/mt7996/mt7996.h b/mt7996/mt7996.h
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -1007,7 +1007,7 @@
 			     struct ieee80211_vif *vif, u8 link_id,
 			     bool changed);
 int mt7996_set_channel(struct mt76_phy *mphy);
-int mt7996_mcu_set_chan_info(struct mt7996_phy *phy, u16 tag);
+int mt7996_mcu_set_chan_info(struct mt7996_phy *phy, u16 tag, bool sta);
 int mt7996_mcu_set_tx(struct mt7996_dev *dev, struct ieee80211_vif *vif,
 		      struct ieee80211_bss_conf *link_conf);
 int mt7996_mcu_set_fixed_rate_ctrl(struct mt7996_dev *dev,
diff --git a/mt7996/testmode.c b/mt7996/testmode.c
--- a/mt7996/testmode.c
+++ b/mt7996/testmode.c
@@ -555,7 +555,7 @@
 
 		/* set channel switch reason */
 		mphy->hw->conf.flags |= IEEE80211_CONF_OFFCHANNEL;
-		mt7996_mcu_set_chan_info(phy, UNI_CHANNEL_SWITCH);
+		mt7996_mcu_set_chan_info(phy, UNI_CHANNEL_SWITCH, false);
 
 		ret = mt76_mcu_send_msg(&dev->mt76, MCU_WM_UNI_CMD(TESTMODE_CTRL), req,
 					sizeof(*req), false);
@@ -569,7 +569,7 @@
 	mphy->hw->conf.flags &= ~IEEE80211_CONF_OFFCHANNEL;
 	memcpy(chandef, &chandef_backup, sizeof(struct cfg80211_chan_def));
 	memcpy(chandef->chan, &chan_backup, sizeof(struct ieee80211_channel));
-	mt7996_mcu_set_chan_info(phy, UNI_CHANNEL_SWITCH);
+	mt7996_mcu_set_chan_info(phy, UNI_CHANNEL_SWITCH, false);
 
 	return ret;
 }
