diff --git a/mt76_connac_mcu.h b/mt76_connac_mcu.h
--- a/mt76_connac_mcu.h
+++ b/mt76_connac_mcu.h
@@ -1089,6 +1089,7 @@
 	MCU_UNI_EVENT_TESTMODE_CTRL = 0x46,
 	MCU_UNI_EVENT_CSI_REPORT = 0x4A,
 	MCU_UNI_EVENT_WED_RRO = 0x57,
+	MCU_UNI_EVENT_PP = 0x5a,
 	MCU_UNI_EVENT_PER_STA_INFO = 0x6d,
 	MCU_UNI_EVENT_ALL_STA_INFO = 0x6e,
 	MCU_UNI_EVENT_SDO = 0x83,
diff --git a/mt7996/main.c b/mt7996/main.c
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -1344,6 +1344,12 @@
 	mt7996_mac_wtbl_update(dev, idx, MT_WTBL_UPDATE_ADM_COUNT_CLEAR);
 	mt7996_mcu_add_sta(dev, link_conf, link_sta, link, msta_link,
 			   CONN_STATE_DISCONNECT, true);
+
+	if (link_sta->eht_cap.has_eht && link_conf->vif->type == NL80211_IFTYPE_STATION) {
+		ret = mt7996_mcu_set_pp_sta_dscb(link->phy, &link_conf->chanreq.oper, link->mt76.omac_idx);
+		if (ret)
+			goto error;
+	}
 
 	rcu_assign_pointer(dev->mt76.wcid[idx], &msta_link->wcid);
 	mt76_wcid_init(&msta_link->wcid, phy->mt76->band_idx);
diff --git a/mt7996/mcu.c b/mt7996/mcu.c
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -1439,6 +1439,39 @@
 }
 
 static void
+mt7996_mcu_pp_event(struct mt7996_dev *dev, struct sk_buff *skb)
+{
+	struct mt7996_mcu_pp_basic_event *event;
+	struct mt7996_mcu_pp_dscb_event *dscb_event;
+	struct mt7996_phy *phy;
+	struct mt76_phy *mphy;
+	u16 report_bitmap;
+
+	event = (struct mt7996_mcu_pp_basic_event *)skb->data;
+
+	switch (le16_to_cpu(event->tag)) {
+	case UNI_EVENT_STATIC_PP_TAG_CSA_DSCB_IE:
+	case UNI_EVENT_STATIC_PP_TAG_DSCB_IE:
+		if (!mt7996_band_valid(dev, event->band_idx))
+			return;
+
+		mphy = mt76_dev_phy(&dev->mt76, event->band_idx);
+		phy = mphy->priv;
+
+		dscb_event = (struct mt7996_mcu_pp_dscb_event *)event;
+		report_bitmap = le16_to_cpu(dscb_event->punct_bitmap);
+
+		if (phy->punct_bitmap == report_bitmap)
+			return;
+
+		if (phy->pp_mode == PP_FW_MODE)
+			phy->punct_bitmap = report_bitmap;
+
+		break;
+	}
+}
+
+static void
 mt7996_mcu_uni_rx_unsolicited_event(struct mt7996_dev *dev, struct sk_buff *skb)
 {
 	struct mt7996_mcu_rxd *rxd = (struct mt7996_mcu_rxd *)skb->data;
@@ -1465,6 +1498,9 @@
 		break;
 	case MCU_UNI_EVENT_PS_SYNC:
 		mt7996_mcu_rx_ps_sync(dev, skb);
+		break;
+	case MCU_UNI_EVENT_PP:
+		mt7996_mcu_pp_event(dev, skb);
 		break;
 #ifdef CONFIG_MTK_DEBUG
 	case MCU_UNI_EVENT_SR:
@@ -6631,7 +6667,8 @@
 		u8 force_bitmap_ctrl;
 		u8 auto_mode;
 		__le16 bitmap;
-		u8 _rsv2[2];
+		u8 csa_enable;
+		u8 _rsv2;
 	} __packed req = {
 		.tag = cpu_to_le16(UNI_CMD_PP_EN_CTRL),
 		.len = cpu_to_le16(sizeof(req) - 4),
@@ -6641,6 +6678,7 @@
 		.force_bitmap_ctrl = (mode == PP_USR_MODE) ? 2 : 0,
 		.auto_mode = pp_auto,
 		.bitmap = cpu_to_le16(bitmap),
+		.csa_enable = false,
 	};
 
 	if (phy->mt76->chanctx->chandef.chan->band == NL80211_BAND_2GHZ ||
@@ -6652,6 +6690,73 @@
 
 	phy->punct_bitmap = bitmap;
 	phy->pp_mode = mode;
+
+	return mt76_mcu_send_msg(&dev->mt76, MCU_WM_UNI_CMD(PP),
+				 &req, sizeof(req), false);
+}
+
+int mt7996_mcu_set_pp_sta_dscb(struct mt7996_phy *phy,
+			       struct cfg80211_chan_def *chandef,
+			       u8 omac_idx)
+{
+	struct mt7996_dev *dev = phy->dev;
+	struct {
+		u8 _rsv1[4];
+
+		__le16 tag;
+		__le16 len;
+		u8 band_idx;
+		u8 omac_idx;
+		u8 eht_op_present;
+		u8 dscb_present;
+		__le16 dscb;
+		u8 ctrl;
+		u8 ccfs0;
+		u8 ccfs1;
+		u8 rsv2[3];
+	} __packed req = {
+		.tag = cpu_to_le16(UNI_CMD_PP_DSCB_CTRL),
+		.len = cpu_to_le16(sizeof(req) - 4),
+
+		.band_idx = phy->mt76->band_idx,
+		.omac_idx = omac_idx,
+		.eht_op_present = true,
+		.dscb_present = !!chandef->punctured,
+		.dscb = cpu_to_le16(chandef->punctured),
+		.ctrl = 0,
+		.ccfs0 = ieee80211_frequency_to_channel(chandef->center_freq1),
+		.ccfs1 = ieee80211_frequency_to_channel(chandef->center_freq1),
+	};
+
+	if (phy->mt76->chanctx->chandef.chan->band == NL80211_BAND_2GHZ ||
+	    phy->punct_bitmap == chandef->punctured)
+		return 0;
+
+	switch (chandef->width) {
+	case NL80211_CHAN_WIDTH_320:
+		req.ctrl |= IEEE80211_EHT_OPER_CHAN_WIDTH_320MHZ;
+		if (chandef->chan->hw_value < req.ccfs1)
+			req.ccfs0 -= 16;
+		else
+			req.ccfs0 += 16;
+		break;
+	case NL80211_CHAN_WIDTH_160:
+		req.ctrl |= IEEE80211_EHT_OPER_CHAN_WIDTH_160MHZ;
+		if (chandef->chan->hw_value < req.ccfs1)
+			req.ccfs0 -= 8;
+		else
+			req.ccfs0 += 8;
+		break;
+	case NL80211_CHAN_WIDTH_80:
+		req.ctrl |= IEEE80211_EHT_OPER_CHAN_WIDTH_80MHZ;
+		req.ccfs0 = 0;
+		break;
+	default:
+		return 0;
+		break;
+	}
+
+	phy->punct_bitmap = cpu_to_le16(chandef->punctured);
 
 	return mt76_mcu_send_msg(&dev->mt76, MCU_WM_UNI_CMD(PP),
 				 &req, sizeof(req), false);
diff --git a/mt7996/mcu.h b/mt7996/mcu.h
--- a/mt7996/mcu.h
+++ b/mt7996/mcu.h
@@ -1292,6 +1292,7 @@
 
 enum {
 	UNI_CMD_PP_EN_CTRL,
+	UNI_CMD_PP_DSCB_CTRL,
 };
 
 enum pp_mode {
@@ -1299,6 +1300,39 @@
 	PP_FW_MODE,
 	PP_USR_MODE,
 };
+
+enum {
+	UNI_EVENT_PP_TAG_ALG_CTRL = 1,
+	UNI_EVENT_STATIC_PP_TAG_DSCB_IE,
+	UNI_EVENT_STATIC_PP_TAG_CSA_DSCB_IE,
+	UNI_EVENT_PP_SHOW_INFO,
+};
+
+struct mt7996_mcu_pp_basic_event {
+	struct mt7996_mcu_rxd rxd;
+
+	u8 __rsv1[4];
+
+	__le16 tag;
+	__le16 len;
+	u8 band_idx;
+	u8 __rsv2[3];
+} __packed;
+
+struct mt7996_mcu_pp_dscb_event {
+	struct mt7996_mcu_rxd rxd;
+
+	u8 __rsv1[4];
+
+	__le16 tag;
+	__le16 len;
+	u8 band_idx;
+	u8 omac_idx;
+	u8 new_dscb;
+	u8 __rsv2;
+	__le16 punct_bitmap;
+	u8 __rsv3[2];
+} __packed;
 
 enum {
 	UNI_CMD_SCS_SEND_DATA,
