diff --git a/mt7996/mac.c b/mt7996/mac.c
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -2685,6 +2685,8 @@
 	mt7996_for_each_phy(dev, phy)
 		clear_bit(MT76_RESET, &phy->mt76->state);
 
+#if LINUX_VERSION_IS_LESS(6,6,0)
+	/* Todo: Check wether it is necessary in kernel 6.6 */
 	mt76_for_each_q_rx(&dev->mt76, i) {
 		if (mtk_wed_device_active(&dev->mt76.mmio.wed) &&
 		    mt76_queue_is_wed_rro(&dev->mt76.q_rx[i]))
diff --git a/mt7996/main.c b/mt7996/main.c
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -2703,6 +2703,7 @@
 	path->mtk_wdma.bss = link->mt76.idx;
 	path->mtk_wdma.queue = 0;
 	path->mtk_wdma.wcid = msta_link->wcid.idx;
+	path->mtk_wdma.tid = mvif->qos_map[dscp];
 
 	if (ieee80211_hw_check(hw, SUPPORTS_AMSDU_IN_AMPDU) &&
 	    mtk_wed_is_amsdu_supported(wed))
@@ -2710,9 +2711,6 @@
 	else
 		path->mtk_wdma.amsdu = 0;
 	ctx->dev = NULL;
-
-	if (path->mtk_wdma.amsdu)
-		path->mtk_wdma.tid = mvif->qos_map[dscp];
 
 	return 0;
 }
diff --git a/mt7996/mmio.c b/mt7996/mmio.c
--- a/mt7996/mmio.c
+++ b/mt7996/mmio.c
@@ -535,6 +535,7 @@
 		}
 
 		wed->wlan.wpdma_rx_glo = wed->wlan.phy_base + hif1_ofs + MT_WFDMA0_GLO_CFG;
+
 		wed->wlan.wpdma_rx[0] = wed->wlan.phy_base + hif1_ofs +
 					MT_RXQ_RING_BASE(MT7996_RXQ_BAND2) +
 					MT7996_RXQ_BAND2 * MT_RING_SIZE;
@@ -565,6 +566,7 @@
 			wed->wlan.wpdma_rx_rro[1] = wed->wlan.phy_base + hif1_ofs +
 						    MT_RXQ_RING_BASE(MT7996_RXQ_RRO_BAND1) +
 						    MT7996_RXQ_RRO_BAND1 * MT_RING_SIZE;
+
 			wed->wlan.wpdma_rx[1] = wed->wlan.phy_base + hif1_ofs +
 						MT_RXQ_RING_BASE(MT7996_RXQ_BAND1) +
 						MT7996_RXQ_BAND1 * MT_RING_SIZE;
diff --git a/wed.c b/wed.c
--- a/wed.c
+++ b/wed.c
@@ -26,6 +26,8 @@
 
 	mt76_free_pending_rxwi(dev);
 
+#if LINUX_VERSION_IS_LESS(6,6,0)
+	/* Todo: Check wether it is necessary in kernel 6.6 */
 	mt76_for_each_q_rx(dev, i) {
 		struct mt76_queue *q = &dev->q_rx[i];
 
@@ -45,6 +47,7 @@
 	page = virt_to_page(wed->rx_buf_ring.rx_page.va);
 	__page_frag_cache_drain(page, wed->rx_buf_ring.rx_page.pagecnt_bias);
 	memset(&wed->rx_buf_ring.rx_page, 0, sizeof(wed->rx_buf_ring.rx_page));
+#endif
 }
 EXPORT_SYMBOL_GPL(mt76_wed_release_rx_buf);
 
