diff --git a/mt76_connac_mcu.h b/mt76_connac_mcu.h
--- a/mt76_connac_mcu.h
+++ b/mt76_connac_mcu.h
@@ -1173,6 +1173,9 @@
 	EE_MODE_EFUSE,
 	EE_MODE_BUFFER,
 };
+
+#define EE_MODE					BIT(0)
+#define EE_PATCH_BACK				BIT(1)
 
 enum {
 	EE_FORMAT_BIN,
diff --git a/mt7996/mcu.c b/mt7996/mcu.c
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -5911,7 +5911,7 @@
 #define PER_PAGE_SIZE		0x400
 	struct mt7996_mcu_eeprom_update req = {
 		.tag = cpu_to_le16(UNI_EFUSE_BUFFER_MODE),
-		.buffer_mode = EE_MODE_BUFFER
+		.buffer_mode = EE_MODE_BUFFER | EE_PATCH_BACK,
 	};
 	u16 eeprom_size = MT7996_EEPROM_SIZE;
 	u8 total = DIV_ROUND_UP(eeprom_size, PER_PAGE_SIZE);
diff --git a/mt7996/mcu.h b/mt7996/mcu.h
--- a/mt7996/mcu.h
+++ b/mt7996/mcu.h
@@ -164,7 +164,7 @@
 
 	__le16 tag;
 	__le16 len;
-	u8 buffer_mode;
+	u8 buffer_mode;		/* bit 0: efuse or buffer mode, bit 1: has patch back */
 	u8 format;
 	__le16 buf_len;
 } __packed;
@@ -214,6 +214,16 @@
 	__le32 addr;
 	__le32 valid;
 	u8 data[MT7996_EEPROM_BLOCK_SIZE];
+} __packed;
+
+struct mt7996_mcu_eeprom_patch {
+	__le16 tag;
+	__le16 len;
+	__le16 adie_offset;
+	__le16 eep_offset;
+	__le16 count;
+	bool complete;
+	u8 rsv;
 } __packed;
 
 struct mt7996_mcu_phy_rx_info {
