diff --git a/mt7996/debugfs.c b/mt7996/debugfs.c
--- a/mt7996/debugfs.c
+++ b/mt7996/debugfs.c
@@ -1185,7 +1185,7 @@
 	if (phy->mt76->cap.has_5ghz) {
 		debugfs_create_u32("dfs_hw_pattern", 0400, dir,
 				   &dev->hw_pattern);
-		debugfs_create_file("radar_trigger", 0200, dir, dev,
+		debugfs_create_file("radar_trigger", 0200, dir, phy,
 				    &fops_radar_trigger);
 		debugfs_create_devm_seqfile(dev->mt76.dev, "rdd_monitor", dir,
 					    mt7996_rdd_monitor);
diff --git a/mt7996/mcu.c b/mt7996/mcu.c
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -6999,7 +6999,7 @@
 		.tag = cpu_to_le16(UNI_RDD_CTRL_PARM),
 		.len = cpu_to_le16(sizeof(req) - 4),
 		.ctrl = RDD_DISABLE_ZW_TIMER,
-		.rdd_idx = MT_RX_SEL2,
+		.rdd_idx = MT_RDD_IDX_BACKGROUND,
 		.disable_timer = disable_timer,
 	};
 
