diff --git a/package/mtk/applications/luci-app-mtwifi-cfg/root/usr/share/luci-app-mtwifi-cfg/wireless-mtk.js b/package/mtk/applications/luci-app-mtwifi-cfg/root/usr/share/luci-app-mtwifi-cfg/wireless-mtk.js
index 20d2f90..7888734 100644
--- a/package/mtk/applications/luci-app-mtwifi-cfg/root/usr/share/luci-app-mtwifi-cfg/wireless-mtk.js
+++ b/package/mtk/applications/luci-app-mtwifi-cfg/root/usr/share/luci-app-mtwifi-cfg/wireless-mtk.js
@@ -94,6 +94,7 @@ function render_signal_badge(signalPercent, signalValue, noiseValue, wrap, mode)

 	return E('div', {
 		'class': wrap ? 'center' : 'ifacebadge',
+		'title': title,
 		'data-signal': signalValue,
 		'data-noise': noiseValue
 	}, [
